tests/platform_tests/test_memory_exhaustion.py

```diff
@@ -13,48 +14,79 @@
+                logging.error("No PDU controller for {}, failed to recover DUT!".format(hostname))
+                return
+            self.pdu_reboot(pdu_controller)
+            # Waiting for SSH connection startup
+            pytest_assert(self.check_ssh_state(localhost, dut_ip, SSH_STATE_STARTED, SSH_STARTUP_TIMEOUT),
+                          'Recover {} by PDU reboot failed'.format(hostname))
+            # Wait until all critical processes are healthy.
+            wait_critical_processes(duthost)
+
+    def test_memory_exhaustion(self, duthost, localhost):
+        dut_ip = duthost.mgmt_ip
+        hostname = duthost.hostname
+        dut_datetime = duthost.get_now_time()
+
+        # Use `tail /dev/zero` to run out of memory completely. Since this command will cause the
+        # DUT reboot, we need to run it in the background (using &) to avoid pytest getting stuck.
+        # We also need to add `nohup` to protect it.
+        cmd = 'nohup tail /dev/zero &'
+        res = duthost.shell(cmd)
+        if not res.is_successful:
+            pytest.fail('DUT {} run command {} failed'.format(hostname, cmd))
+
+        # Waiting for SSH connection shutdown
+        pytest_assert(self.check_ssh_state(localhost, dut_ip, SSH_STATE_ABSENT, SSH_SHUTDOWN_TIMEOUT),
+                      'DUT {} did not shutdown'.format(hostname))
+        # Waiting for SSH connection startup
+        pytest_assert(self.check_ssh_state(localhost, dut_ip, SSH_STATE_STARTED, SSH_STARTUP_TIMEOUT),
+                      'DUT {} did not startup'.format(hostname))
+        # Wait until all critical processes are healthy.
+        wait_critical_processes(duthost)
+        # Verify DUT uptime is later than the time when the test case started running.
+        dut_uptime = duthost.get_up_time()
+        pytest_assert(dut_uptime > dut_datetime, "Device {} did not reboot".format(hostname))
+
+    def check_ssh_state(self, localhost, dut_ip, expected_state, timeout=60):
+        """
+        Check the SSH state of DUT.
+
+        :param localhost: A `tests.common.devices.local.Localhost` Object.
+        :param dut_ip: A string, the IP address of DUT.
+        :param expected_state: A string, the expected SSH state.
+        :param timeout: An integer, the maximum number of seconds to wait for.
+        :return: A boolean, True if SSH state is the same as expected
+                          , False otherwise.
+        """
+        res = localhost.wait_for(host=dut_ip,
+                                 port=SONIC_SSH_PORT,
+                                 state=expected_state,
+                                 search_regex=SONIC_SSH_REGEX,
+                                 delay=10,
+                                 timeout=timeout,
+                                 module_ignore_errors=True)
+        return not res.is_failed and 'Timeout' not in res.get('msg', '')
+
+    def pdu_reboot(self, pdu_controller):
+        pdu_controller.turn_off_outlet()
```

Comment:
what happens if the pdu controller doesn't exists or the control command failed?

Reply:
I checked if the PDU controller exists at line 34. I added some code to check the results of these control commands in a new commit.